### PR TITLE
Creation of Battle Mode for CLI Integration

### PR DESCRIPTION
**Pull request by jintae412**
We modified the mode.c file to allow for a Battle Mode to run battles for the CLI, and then made an example of how this would work using a monkey-patched Fight command. This required us to modify the ctx struct to have a field for a battle context so that the CLI would have a way to access a given battle, and modify/create the relevant functions. In addition, modifications to the code were necessary so that actions produced the necessary strings that would be sent to the CLI to be printed through the CLI print functions. Battle_items and move structures must have their "name" and "info" fields respectively now in all-caps to ensure compatability with the parsed inputs from the CLI, which automatically makes all of the inputs all-caps.

### CODE REVIEW

File: include/game-state/game.h [hunk 1]
```diff
@@ -78,6 +79,9 @@ typedef struct game {
 
     /* time when game started */
     //int time_start;
+
+    /* pointer to the current battle context */
+    chiventure_ctx_battle_t *battle_ctx;
```
Comment (borjasotomayor) [inline include/game-state/game.h:84]: Rename `chiventure_ctx_battle_t` to `battle_ctx_t` for coherence with other struct names (note how no other struct in `game_t` -or in the entire code- uses the prefix `chiventure_ctx`).

**Review by borjasotomayor — CHANGES_REQUESTED** [review]
The changes to the game state code look good to me, but there is a small naming change that has to be made.

Additionally, someone should create a backlog item for moving "modes" out of Game State, since they make more sense inside CLI (however, that is a much larger task, and one that can simply be added to the backlog)

**Review by lczielinski — APPROVED** [review]
Wow, this looks awesome! I think there's a typo on one of the header comments but that's about it

File: include/battle/battle_print.h [hunk 2]
```diff
@@ -70,4 +70,33 @@ char *print_start_battle(battle_t *b);
...
+ */
+char *print_start_turn(battle_t* b);
+
+/* Returns a string for the avaliable battle_items for the player
+ * Parameter:
+ *  - b = pointer to the battle
+ * Returns:
+ *  A string containing the available items
+ */ 
+int *print_battle_items(battle_t *b, char *string);
```
[the remark this answers is not in the document]
Comment (jintae412) [reply include/battle/battle_print.h:94]: Yes, this has been fixed. Thanks!

File: include/game-state/game.h (hunk 1 above)
Comment (jintae412) [reply include/game-state/game.h]: All instances of the chiventure_ctx_battle_t struct should have been renamed to battle_ctx_t.

**Review by borjasotomayor — APPROVED** [review]

**Comment by Elsa-Athiley** [comment]
Just pushed some minor changes, it seems like some code was refactored so that the `chiventure_ctx_t` struct now has a field for `cli_ctx_t` which now holds the lookup table so I changed the way we were accessing it so it no longer caused an error. I also added another case so if an invalid command is passed in for the 'list item/move' it returns a message saying ivalid command. Also fixed a small typo.

File: src/battle/src/battle_flow.c [hunk 3]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+    char *string = print_battle_move(b, b->turn, move);
 
     if(battle_over(b) == BATTLE_VICTOR_PLAYER)
     {
         /* print stub: should tel player they won */
         award_xp(b->player->stats, 2.0);
         ctx->status = BATTLE_VICTOR_PLAYER;
+    }
+    
+    if(battle_over(b) == BATTLE_IN_PROGRESS){
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:127]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 4]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+{
+    battle_t *b = ctx->game->battle;
+    char *string = calloc(150, sizeof(char));
+
+    if (ctx == NULL)
+    {
+        snprintf(string, 150, "FAILURE\n");
+        return string;
+    }
+    if (item == NULL){
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:146]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 5]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+        snprintf(string, 150, "FAILURE\n");
+        return string;
+    }
+    if (item->quantity <= 0){
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:150]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 6]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+    snprintf(string, 150, "You used the %s\n", item_name);
 
-        return SUCCESS;
+    if (usage == FAILURE) 
+    {
+        snprintf(string, 150, "That item is Unavailable.\n");
+        return string;
+    } 
+
+    if(battle_over(b) == BATTLE_IN_PROGRESS){
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:167]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 7]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
 
-    int res = enemy_make_move(ctx);
+    return string;
+}
 
-    return res;
+/* see battle_flow.h */
+char *battle_flow_list(battle_ctx_t *ctx, char* label)
+{
+    if (strcmp(label, "ITEMS") == 0) {
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:179]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 8]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+        battle_t *b = ctx->game->battle;
+
+        char *string = calloc(500, sizeof(char));
+        print_battle_items(b, string);
+
+        return string;
+    } if (strcmp(label, "MOVES") == 0) {
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:186]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_flow.c [hunk 9]
```diff
@@ -114,42 +115,117 @@ int battle_flow(chiventure_ctx_battle_t *ctx, move_t *move, char* target)
...
+        battle_t *b = ctx->game->battle;
+
+        char *string = calloc(500, sizeof(char));
+        print_moves(b, string);
+
+        return string;
+    } else {
```
Comment (firatciftci) [inline src/battle/src/battle_flow.c:193]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of this line to a new line below.

File: src/battle/src/battle_logic.c [hunk 10]
```diff
@@ -50,13 +51,30 @@ turn_t goes_first(battle_t *b)
...
+
+/* see battle_logic.h */
+battle_item_t *find_battle_item(battle_item_t *inventory, char *input)
 {
     battle_item_t *temp;
 
     DL_FOREACH(inventory, temp)
-    {
-        if (temp->id == id)
+    {
```
Comment (firatciftci) [inline src/battle/src/battle_logic.c:76]: GitHub shows extra white space being added to this line which I assume should not be there.

File: src/battle/src/battle_print.c [hunk 11]
```diff
@@ -102,3 +96,82 @@ char *print_battle_winner(battle_status_t status, int xp)
...
+
+    char *string = calloc(BATTLE_BUFFER_SIZE + 1, sizeof(char));
+
+
+
+    snprintf(string, BATTLE_BUFFER_SIZE, "\nTo use a Move, type 'Use [insert move name here]'\n\n"
+                                         "To use an Item, type 'Consume [insert item name here]'\n\n"
+                                         "To see the list available moves, type 'List Moves'\n\n"
+                                         "To see the list available items, type 'List Items'\n");
+    //print_moves(b,string);
```
Comment (firatciftci) [inline src/battle/src/battle_print.c:115]: Make sure to remove any unused commented out code.

File: src/game-state/src/mode.c [hunk 12]
```diff
@@ -134,3 +137,62 @@ int run_conversation_mode(char *input, cli_callback callback_func,
     return SUCCESS;
 }
 
+/* see mode.h */
+int run_battle_mode (char *input, cli_callback callback_func, 
+                          void *callback_args, chiventure_ctx_t *ctx)
+{
+    char **parsed_input = parse(input);
+    if (parsed_input == NULL)
+    {
```
Comment (firatciftci) [inline src/game-state/src/mode.c:146]: This file seems to be using the end-of-line approach for opening curly brackets, so I think you should move the `{` on this line to the end of the line just above.

File: src/game-state/src/mode.c [hunk 13]
```diff
@@ -134,3 +137,62 @@ int run_conversation_mode(char *input, cli_callback callback_func,
...
+
+        if (move == NULL) {
+            return callback_func(ctx, "That Move does not exist.", callback_args);
+        }
+
+        string = battle_flow_move(battle_ctx, move, battle_ctx->game->battle->enemy->name);
+
+    } else if (strcmp(parsed_input[0], "CONSUME") == 0) {
+        char *stringed_input = calloc(100, sizeof(char));
+        for(int i = 1; parsed_input[i] != NULL; i++) {
```
Comment (firatciftci) [inline src/game-state/src/mode.c:166]: Make sure to place a single space character between `for` and `(`.

**Review by firatciftci — CHANGES_REQUESTED** [review]
I've requested some minor stylistic changes.

**Review by firatciftci — APPROVED** [review]
Looks good to me!

**Comment by jintae412** [comment]
To summarize this PR, a mode was created for conducting battles through the CLI. The run_battle_mode function uses if/else statements to go through the potential inputs given and so run the relevant battle_flow and battle_logic functions. This required liberal use of statements to print to the CLI as well as easy access to the required strings to output, so numerous changes to the battle_print functions as well as battle_flow and battle_logic were necessary. Similarly, keeping in line with how parsing of CLI input is conducted, as of right now items and moves work only should their names be in all lowercase letters.

The example created for the demo and to showcase this feature required a monkeypatched "FIGHT" command, given that the CLI team has not yet officially created one yet. Furthermore, there remains yet a bug in which one of the possible consumable items created by the default item creation function, the "elixir of life" occasionally causes a segfault while running the example. It is unknown why this is the case, but a backlog issue making note of it has been created.

**Comment by firatciftci** [comment]
Issue Score: ✔️++

Comments:
Good work!

**Pull request merged**